### PR TITLE
[kernel] Add triton kernel for context attention without padding

### PR DESCRIPTION
**Pull request by yuanheng-zhao**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

Implemented unpadded context attention in #5193 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Add triton kernel for non-padding context attention. Use flash attention V2 implementation based on "FlashAttention-2: Faster Attention with Better Parallelism and Work Partitioning" by Tri Dao, 2023 and modified from triton tutorial https://triton-lang.org/main/getting-started/tutorials/06-fused-attention.html

image

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: tests/test_infer_ops/triton/test_context_attn_unpad.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    dtype = torch.float16
+    device = get_current_device()
+    num_seqs = bsz
+    num_heads = 16
+    # Consider MHA for now and thus num_kv_heads == num_heads
+    num_kv_heads = num_heads
+    head_size = 128
+    # block_size = 8
+    # max_num_blocks_per_seq = 10
```
Comment (yuanheng-zhao) [inline tests/test_infer_ops/triton/test_context_attn_unpad.py:79]: To be updated in the preceding PR

**Review by FrankLeeeee — APPROVED** [review]
LGTM.

File: tests/test_infer_ops/triton/test_context_attn_unpad.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
+import pytest
+import torch
+import torch.nn.functional as F
+from packaging import version
+
+from colossalai.kernel.triton import context_attention_unpadded
+from colossalai.utils import get_current_device
+
+try:
```
Comment (CjhHa1) [inline tests/test_infer_ops/triton/test_context_attn_unpad.py:9]: this is always true?

Comment (FrankLeeeee) [reply tests/test_infer_ops/triton/test_context_attn_unpad.py:9]: I think it might be removed by pre-commit

Comment (FrankLeeeee) [reply tests/test_infer_ops/triton/test_context_attn_unpad.py:9]: Try add `# noqa` to the import statement.

Comment (yuanheng-zhao) [reply tests/test_infer_ops/triton/test_context_attn_unpad.py:9]: Import statement revised.
